<a href="https://colab.research.google.com/github/keinam53/Kurs_Pandas/blob/master/Uczenie_nadzorowane/2_Regresja/1_Regresja_liniowa_rownanie_normalne.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

###Regresja liniowa - równanie normalne
1. [Import bibliotek](#0)
2. [Przygotowanie danych](#1)
3. [Równanie normalne](#2)
4. [Regresja liniowa przy pomocy scikit-learn](#3)

### <a name='0'></a> Import bibliotek

In [1]:
import numpy as np

### <a name='1'></a> Przygotowanie danych

In [3]:
X1 = np.array([1, 2, 3, 4, 5, 6])
Y = np.array([3000, 3250, 3500, 3750, 4000, 4250])
m = len(X1)

print(f'Lata pracy: {X1}')
print(f'Wynagrodzenie: {Y}')
print(f'Liczba próbek: {m}')

Lata pracy: [1 2 3 4 5 6]
Wynagrodzenie: [3000 3250 3500 3750 4000 4250]
Liczba próbek: 6


In [4]:
#zmiana kształtu danych - 6 wierszy, 1 kolumna
X1 = X1.reshape(m, 1)
print(X1)
print(f'\n{X1.shape}')

[[1]
 [2]
 [3]
 [4]
 [5]
 [6]]

(6, 1)


In [6]:
#wygenerowanie wektora jedynek - bias
bias = np.ones((m, 1))
print(bias)
print(f'\n{bias.shape}')

[[1.]
 [1.]
 [1.]
 [1.]
 [1.]
 [1.]]

(6, 1)


In [15]:
#dodanie X1 i bias
X = np.append(bias, X1, axis=1)
print(X)
print(f'\n{X1.shape}')

[[1. 1.]
 [1. 2.]
 [1. 3.]
 [1. 4.]
 [1. 5.]
 [1. 6.]]

(6, 1)


### <a name='2'></a> Równanie normalne

Regresja liniowa w $R^2$:
$$Y = w_0 + w_1X_1$$
$$Y = W^TX$$
gdzie: $$W= \left[\begin{matrix}w_0\\w_1\end{matrix}\right]$$ $$ X= \left[\begin{matrix}1\\X_1\end{matrix}\right] $$ stąd $$ W^T= \left[\begin{matrix}w_0&w_1\end{matrix}\right] $$    
$$Y = W^TX = \left[\begin{matrix}w_0&w_1\end{matrix}\right] \cdot  \left[\begin{matrix}1\\X_1\end{matrix}\right] = w_0 + w_1X_1$$

Równanie normalne - równanie pozwalające obliczyć minimum funkcji straty (o ile istnieje)
$$W = (X^TX)^{-1}(X^TY)$$

In [16]:
#wykonanie ręcznych obliczeń z wzoru na W
#pierwszy nawias
np.dot(X.transpose(), X)

array([[ 6., 21.],
       [21., 91.]])

In [18]:
#macierz odwrotna
L = np.linalg.inv(np.dot(X.transpose(), X))
L

array([[ 0.86666667, -0.2       ],
       [-0.2       ,  0.05714286]])

In [19]:
#drugi nawias
P = np.dot(X.transpose(), Y)
P

array([21750., 80500.])

In [20]:
#Wynik
np.dot(L, P)

array([2750.,  250.])

Końcowa postać modelu$$Y = 2750 + 250X_1$$

### <a name='3'></a> Regresja liniowa przy pomocy scikit-learn

In [23]:
from sklearn.linear_model import LinearRegression

regression = LinearRegression()
regression.fit(X1, Y)

print(regression.intercept_)  #intercept_ - współczynnik w0
print(regression.coef_[0])  #coef_[0] - współczynnik w1

2750.0
250.0
